File: backend/app/schemas/course.py
from __future__ import annotations

import uuid
from typing import Any

from pydantic import BaseModel

from app.schemas.task_pack import TaskPackSummary


class CourseSessionBrief(BaseModel):
    id: uuid.UUID
    number: int
    title: str
    duration: str

    model_config = {"from_attributes": True}


class CourseModuleRead(BaseModel):
    id: uuid.UUID
    number: int
    title: str
    overview_md: str
    prerequisites: str | None
    estimated_time: str
    skills_json: list[str]
    bridge_text: str | None
    session_count: int
    sessions: list[CourseSessionBrief]

    model_config = {"from_attributes": True}


class CourseModuleBrief(BaseModel):
    id: uuid.UUID
    number: int
    title: str
    estimated_time: str
    session_count: int

    model_config = {"from_attributes": True}


class CourseSessionRead(BaseModel):
    id: uuid.UUID
    module_id: uuid.UUID
    number: int
    title: str
    overview_md: str
    rubric_md: str | None
    questions_md: str | None
    exercises_md: str | None
    ai_tasks_md: str | None
    resources_md: str | None
    duration: str
    objectives_json: list[str]
    outline_json: list[dict[str, Any]]
    task_packs: list[TaskPackSummary]
    module: CourseModuleBrief

    model_config = {"from_attributes": True}


class SessionProgressItem(BaseModel):
    session_number: int
    session_title: str
    completed: bool
    packs_total: int
    packs_completed: int


class ModuleProgressItem(BaseModel):
    module_number: int
    module_title: str
    sessions: list[SessionProgressItem]
    completed_sessions: int
    total_sessions: int


class CourseProgressResponse(BaseModel):
    modules: list[ModuleProgressItem]
    total_sessions: int
    completed_sessions: int
    current_session: int | None


class CourseOverviewResponse(BaseModel):
    modules: list[CourseModuleRead]
    total_modules: int
    total_sessions: int

File: frontend/src/types/course.ts
export interface CourseSessionBrief {
  id: string;
  number: number;
  title: string;
  duration: string;
}

export interface CourseModuleRead {
  id: string;
  number: number;
  title: string;
  overview_md: string;
  prerequisites: string | null;
  estimated_time: string;
  skills_json: string[];
  bridge_text: string | null;
  session_count: number;
  sessions: CourseSessionBrief[];
}

export interface TaskPackSummary {
  id: string;
  slug: string;
  name: string;
  description: string;
  language: string;
  task_count: number;
  created_at: string;
}

export interface CourseModuleBrief {
  id: string;
  number: number;
  title: string;
  estimated_time: string;
  session_count: number;
}

export interface CourseSessionRead {
  id: string;
  module_id: string;
  number: number;
  title: string;
  overview_md: string;
  rubric_md: string | null;
  questions_md: string | null;
  exercises_md: string | null;
  ai_tasks_md: string | null;
  resources_md: string | null;
  duration: string;
  objectives_json: string[];
  outline_json: Array<{ title: string; items: string[] }>;
  task_packs: TaskPackSummary[];
  module: CourseModuleBrief;
}

export interface SessionProgressItem {
  session_number: number;
  session_title: string;
  completed: boolean;
  packs_total: number;
  packs_completed: number;
}

export interface ModuleProgressItem {
  module_number: number;
  module_title: string;
  sessions: SessionProgressItem[];
  completed_sessions: number;
  total_sessions: number;
}

export interface CourseProgressResponse {
  modules: ModuleProgressItem[];
  total_sessions: number;
  completed_sessions: number;
  current_session: number | null;
}

export interface CourseOverviewResponse {
  modules: CourseModuleRead[];
  total_modules: number;
  total_sessions: number;
}
